### PR TITLE
Actually use Device Type for mails

### PR DESCRIPTION
**Pull request by dfunkt**
- match Bitwarden behaviour
- remove upcase_first, no longer needed and it would also interfere with device types like iOS, macOS

Examples:

Chrome Extension
Before -> New Device Logged In From/Incomplete Two-Step Login From/Device Type: Chrome
After -> New Device Logged In From/Incomplete Two-Step Login From/Device Type: Chrome Extension

Poco F3 (shows up as M2012K11AG)
Before -> New Device Logged In From/Incomplete Two-Step Login From/Device Type: M2012K11AG
After -> New Device Logged In From/Incomplete Two-Step Login From/Device Type: Android

### CODE REVIEW

**Comment by jjlin** [comment]
If the device type is Android or iOS, the mobile client should include a device model as the name, so the notification email should ideally include that info for those device types. I'm not sure if any other device types include useful model/name info.

**Comment by dfunkt** [comment]
Let's not forget that the email template says "Device Type", not "Device Type and Name/Model" or anything like that.
Example: https://github.com/dani-garcia/vaultwarden/blob/248e561b3fe6a8172751374df980c6cd43c841d5/src/static/templates/email/new_device_logged_in.html.hbs#L22
Ideally I try to stick as close as possible to Bitwarden's behaviour when implementing these changes, and the emails from them don't specify the device name/model number either.

Nonetheless, let's see what others have to say about this change.

**Comment by jjlin** [comment]
Vaultwarden does various things differently from Bitwarden, as long as they are compatible improvements. For example, Bitwarden does not even have a 2FA incomplete notification AFAIK.

**Comment by BlackDex** [comment]
I agree with @jjlin.
For all the client functionality we should stay as close as possible.
Server-side specific changes should be looked at if it has no impact client side or might alter a way of working.

But for something like emails shouldn't be an issue. 
I actually love it shows we the device name and not only type.
So both would be better. 

I even was thinking about adding basic location detection based upon IP.

**Comment by dfunkt** [comment]
New variant with device name and device type included.
Examples:
Screenshot from 2024-09-04 14-44-43
Screenshot from 2024-09-04 14-44-58

File: migrations/mysql/2024-09-04-091351_use_device_types/down.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+ALTER TABLE `twofactor_incomplete` DROP COLUMN `device_atype`;
```
Comment (jjlin) [inline migrations/mysql/2024-09-04-091351_use_device_types/down.sql:1]:
```suggestion
ALTER TABLE `twofactor_incomplete` DROP COLUMN `device_type`;
```
Same name change for all the other migration files

File: src/db/models/two_factor_incomplete.rs [hunk 2]
```diff
@@ -13,6 +13,7 @@ db_object! {
         // must complete 2FA login before being added into the devices table.
         pub device_uuid: String,
         pub device_name: String,
+        pub device_atype: i32,
```
Comment (jjlin) [inline src/db/models/two_factor_incomplete.rs:16]:
```suggestion
        pub device_type: i32,
```

File: src/db/models/two_factor_incomplete.rs [hunk 3]
```diff
@@ -23,6 +24,7 @@ impl TwoFactorIncomplete {
         user_uuid: &str,
         device_uuid: &str,
         device_name: &str,
+        device_atype: &i32,
```
Comment (jjlin) [inline src/db/models/two_factor_incomplete.rs:27]:
```suggestion
        device_type: i32,
```

File: src/db/schemas/mysql/schema.rs [hunk 4]
```diff
@@ -169,6 +169,7 @@ table! {
         user_uuid -> Text,
         device_uuid -> Text,
         device_name -> Text,
+        device_atype -> Integer,
```
Comment (jjlin) [inline src/db/schemas/mysql/schema.rs:172]:
```suggestion
        device_type -> Integer,
```
Same comment for the other schema files

File: src/api/identity.rs [hunk 5]
```diff
@@ -535,7 +551,7 @@ async fn twofactor_auth(
         return Ok(None);
     }
 
-    TwoFactorIncomplete::mark_incomplete(&user.uuid, &device.uuid, &device.name, ip, conn).await?;
+    TwoFactorIncomplete::mark_incomplete(&user.uuid, &device.uuid, &device.name, &device.atype, ip, conn).await?;
```
Comment (jjlin) [inline src/api/identity.rs:554]:
```suggestion
    TwoFactorIncomplete::mark_incomplete(&user.uuid, &device.uuid, &device.name, device.type, ip, conn).await?;
```

**Comment by dfunkt** [comment]
I chose atype on purpose to match how it was in device.rs:
https://github.com/dani-garcia/vaultwarden/blob/248e561b3fe6a8172751374df980c6cd43c841d5/src/db/models/device.rs#L19
Which one do you prefer @BlackDex? atype or just type?

**Comment by BlackDex** [comment]
`type` is the correct name. The only reason there is an `a` in front is because you can't use that as a variable name.

**Comment by jjlin** [comment]
> I chose atype on purpose to match how it was in device.rs:
> 
> https://github.com/dani-garcia/vaultwarden/blob/248e561b3fe6a8172751374df980c6cd43c841d5/src/db/models/device.rs#L19

This is only called `atype` because you can't have an identifier that's just called `type`. There are many other existing instances of `device_type` in the code.

**Comment by dfunkt** [comment]
Hopefully things are good now

File: src/api/identity.rs [hunk 6]
```diff
@@ -265,7 +265,15 @@ async fn _password_login(
     let twofactor_token = twofactor_auth(&user, &data, &mut device, ip, conn).await?;
 
     if CONFIG.mail_enabled() && new_device {
-        if let Err(e) = mail::send_new_device_logged_in(&user.email, &ip.ip.to_string(), &now, &device.name).await {
+        if let Err(e) = mail::send_new_device_logged_in(
+            &user.email,
+            &ip.ip.to_string(),
+            &now,
+            &device.name,
+            &DeviceType::from_i32(device.atype).to_string(),
```
Comment (BlackDex) [inline src/api/identity.rs:273]: I would suggest to just sent `&Device` instead of two different values from the same object.
So `&device` should be enough.

File: src/api/identity.rs [hunk 7]
```diff
@@ -421,7 +429,15 @@ async fn _user_api_key_login(
 
     if CONFIG.mail_enabled() && new_device {
         let now = Utc::now().naive_utc();
-        if let Err(e) = mail::send_new_device_logged_in(&user.email, &ip.ip.to_string(), &now, &device.name).await {
+        if let Err(e) = mail::send_new_device_logged_in(
+            &user.email,
+            &ip.ip.to_string(),
+            &now,
+            &device.name,
+            &DeviceType::from_i32(device.atype).to_string(),
```
Comment (BlackDex) [inline src/api/identity.rs:437]: Same here as above of course

**Review by BlackDex — CHANGES_REQUESTED** [review]

**Comment by dfunkt** [comment]
@BlackDex Not sure if this is what you meant, but that's all I could come up with. I'm not particularly familiar with Rust (or programming in general tbh).

File: src/mail.rs [hunk 8]
```diff
@@ -446,9 +446,10 @@ pub async fn send_invite_confirmed(address: &str, org_name: &str) -> EmptyResult
     send_email(address, &subject, body_html, body_text).await
 }
 
-pub async fn send_new_device_logged_in(address: &str, ip: &str, dt: &NaiveDateTime, device: &str) -> EmptyResult {
+pub async fn send_new_device_logged_in(address: &str, ip: &str, dt: &NaiveDateTime, device: &Device) -> EmptyResult {
     use crate::util::upcase_first;
-    let device = upcase_first(device);
+    let device_name = upcase_first(&device.name);
+    let device_type = DeviceType::from_i32(device.atype).to_string();
```
Comment (BlackDex) [inline src/mail.rs:452]: These do not have to be a variable.
They can be added directly to the `json!()` bellow on lines 461 and 462.

File: src/mail.rs [hunk 9]
```diff
@@ -457,17 +458,24 @@ pub async fn send_new_device_logged_in(address: &str, ip: &str, dt: &NaiveDateTi
...
+pub async fn send_incomplete_2fa_login(
+    address: &str,
+    ip: &str,
+    dt: &NaiveDateTime,
+    device_name: &str,
+    device_type: &str,
+) -> EmptyResult {
     use crate::util::upcase_first;
-    let device = upcase_first(device);
+    let device_name = upcase_first(device_name);
```
Comment (BlackDex) [inline src/mail.rs:478]: Same here, move this directly into the `json!()` below in line 487.

**Review by BlackDex — CHANGES_REQUESTED** [review]
Yes, that is what i mend indeed.

I just noticed something else too. Not that important, but no need to create new variables for those items actually.

File: src/mail.rs (hunk 8 above)
Comment (dfunkt) [reply src/mail.rs]: Done

File: src/mail.rs (hunk 9 above)
Comment (dfunkt) [reply src/mail.rs]: Done

File: src/mail.rs [hunk 10]
```diff
@@ -457,17 +456,23 @@ pub async fn send_new_device_logged_in(address: &str, ip: &str, dt: &NaiveDateTi
             "url": CONFIG.domain(),
             "img_src": CONFIG._smtp_img_src(),
             "ip": ip,
-            "device": device,
```
Comment (dfunkt) [inline src/mail.rs:460]: Maybe I should modify this to ```&device.name``` as well in order to keep existing templates compatible?

File: src/mail.rs [hunk 11]
```diff
@@ -476,7 +481,8 @@ pub async fn send_incomplete_2fa_login(address: &str, ip: &str, dt: &NaiveDateTi
...
```
Comment (dfunkt) [inline src/mail.rs:479]: Same here

File: src/mail.rs (hunk 10 above)
Comment (BlackDex) [reply src/mail.rs:460]: Well, users will then miss the other variable. So they should now be triggered to adjust the templates.
We could add an extra item in the changelogs for this.

I would leave it as  you have right now from my point of view.

**Review by BlackDex — APPROVED** [review]

**Review by dani-garcia — APPROVED** [review]

**Comment by gbe0** [comment]
With this merged I get the following on startup:

```
[2024-09-19 10:17:36.721][panic][ERROR] thread 'main' panicked at 'Error running migrations: QueryError(DieselMigrationName { name: "2024-09-04-091351_use_device_type_for_mails", version: MigrationVersion("20240904091351") }, DatabaseError(Unknown, "syntax error at or near \"`\""))': src/db/mod.rs:495
   0: vaultwarden::init_logging::{{closure}}
   1: std::panicking::rust_panic_with_hook
   2: std::panicking::begin_panic_handler::{{closure}}
   3: std::sys::backtrace::__rust_end_short_backtrace
   4: rust_begin_unwind
   5: core::panicking::panic_fmt
   6: core::result::unwrap_failed
   7: vaultwarden::db::DbPool::from_config
   8: vaultwarden::main::{{closure}}
   9: vaultwarden::main
  10: std::sys::backtrace::__rust_begin_short_backtrace
  11: main
  12: <unknown>
  13: __libc_start_main
  14: _start
```

My DB is `PostgreSQL 16.4 on x86_64-pc-linux-gnu, compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit`.

I rolled back and there are no issues.

**Comment by Gerardv514** [comment]
@dfunkt i just checked this out. I have an iOS device and just tried to login to the web vault using Google chrome iOS app, i closed out on the 2fa page which prompted the 2fa incomplete notification. Device type and device name both say safari, is that normal?

**Comment by dfunkt** [comment]
I think that might be due to Chrome using the browser engine as Safari on iOS. I expect this happens with Bitwarden as well, but I don't have any iOS devices in order to test this theory.

TL;DR: I think this is normal

**Comment by Gerardv514** [comment]
I also attempted the iOS bitwarden app, and device name was iPhone 15,2 and device type was iOS. 

So as you said probably normal due to browser not passing the info along.
